### PR TITLE
feat(e2e): create FeeOracleV2 deployer

### PR DESCRIPTION
**Pull request by Zodomo**
Created a FeeOracleV2 deployment routine that uses Create3. We cannot upgrade FeeOracleV1 as V2 has a different storage layout. This routine does not currently include upgrade logic.

issue: #1951

### CODE REVIEW

File: lib/contracts/feeoraclev2/feeparams.go [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	// used cached pricer, to avoid multiple price requests for same token
+	pricer = tokens.NewCachedPricer(pricer)
+
+	srcChain, ok := evmchain.MetadataByID(srcChainID)
+	if !ok {
+		return nil, errors.New("meta by chain id", "chain_id", srcChainID)
+	}
+
+	var resp []bindings.IFeeOracleV2FeeParams
+	for _, destChainID := range destChainIDs {
```
Comment (kevinhalliday) [inline lib/contracts/feeoraclev2/feeparams.go:29]: We might also want to track srcChainId fee params on the source. Litle weird,  but we do that right now. And I know there was a reason. I'll go find if we still need to

**Review by kevinhalliday — COMMENTED** [review]
Overall looks pretty good.

Couple left of `FeeOracleV1` references in left comments we should update

**Review by chmllr — APPROVED** [review]
Code LGTM, can't really review the semantics yet.

File: e2e/app/feeoraclev2.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+import (
+	"context"
+
+	"github.com/omni-network/omni/lib/contracts/feeoraclev2"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/log"
+)
+
+func DeployFeeOracleV2(ctx context.Context, def Definition) error {
```
Comment (kc1116) [inline e2e/app/feeoraclev2.go:11]:
```suggestion
// DeployFeeOracleV2 deploys fee oracle v2.
func DeployFeeOracleV2(ctx context.Context, def Definition) error {
```

add godocs to exported funcs

File: e2e/app/feeoraclev2.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+func deployFeeOracleV2(ctx context.Context, def Definition) error {
+	_, ok := def.Testnet.OmniEVMChain()
+	if !ok {
+		return errors.New("no omni evm chain")
+	}
+
+	allChains := def.Testnet.EVMChains()
+	for _, chain := range allChains {
+		backends := def.Backends()
```
Comment (kc1116) [inline e2e/app/feeoraclev2.go:27]:
```suggestion
	allChains := def.Testnet.EVMChains()
	backends := def.Backends()
	for _, chain := range allChains {
```
backends initialized on every loop iteration, move outside loop.

File: lib/contracts/feeoraclev2/feeparams.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+		// Add postsTo rates if not already present
+		if destChain.PostsTo != 0 && !contains(resp, destChain.PostsTo) {
+			resp = append(resp, bindings.IFeeOracleV2FeeParams{
+				ChainId:      destChain.PostsTo,
+				ExecGasPrice: uint64(params.GWei),
+				DataGasPrice: uint64(params.GWei),
+				ToNativeRate: rateToNumerator(1), // Stub native rates for now
+			})
+		}
```
[the remark this answers is not in the document]
Comment (Zodomo) [reply lib/contracts/feeoraclev2/feeparams.go:50]: Removed this block

File: lib/contracts/feeoraclev2/feeparams.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		log.Warn(ctx, "Failed fetching gas price, using default 1 Gwei", err, "dest_chain", destChain.Name)
+		gasPrice = big.NewInt(params.GWei)
+	}
+
+	return bindings.IFeeOracleV2FeeParams{
+		ChainId:      destChain.ChainID,
+		ExecGasPrice: withGasPriceShield(float64(gasPrice.Uint64())),
+		DataGasPrice: withGasPriceShield(float64(gasPrice.Uint64())),
+		ToNativeRate: rateToNumerator(toNativeRate),
+	}, nil
```
[the remark this answers is not in the document]
Comment (Zodomo) [reply lib/contracts/feeoraclev2/feeparams.go:83]: Good catch, now retrieving data gas price using SuggestGasPrice on the postsTo chain

File: e2e/app/feeoraclev2.go (hunk 3 above)
Comment (Zodomo) [reply e2e/app/feeoraclev2.go:27]: Moved outside of loop

File: e2e/app/feeoraclev2.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+		destChainIDs := make([]uint64, 0, len(allChains)-1)
+		for _, destChain := range allChains {
+			if destChain.ChainID != chain.ChainID {
+				destChainIDs = append(destChainIDs, destChain.ChainID)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (Zodomo) [reply e2e/app/feeoraclev2.go:34]: Moved to getChainIDs helper function. Also revised to retrieve all chainIDs as we also set FeeParams on source chain as well.

File: e2e/app/feeoraclev2.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+func DeployFeeOracleV2(ctx context.Context, def Definition) error {
+	if err := deployFeeOracleV2(ctx, def); err != nil {
+		return errors.Wrap(err, "deploy fee oracle v2")
+	}
+
+	return nil
+}
+
+func deployFeeOracleV2(ctx context.Context, def Definition) error {
```
[the remark this answers is not in the document]
Comment (Zodomo) [reply e2e/app/feeoraclev2.go:19]: Condensed them

File: lib/contracts/feeoraclev2/feeparams.go (hunk 1 above)
Comment (Zodomo) [reply lib/contracts/feeoraclev2/feeparams.go:29]: All chainIDs are now passed in instead of filtering out the source

**Review by kevinhalliday — APPROVED** [review]

**Pull request merged**